NEXT-13735 Fix DBAL Entity Hydration when one entity is hydrated multiple times with different associations loaded

AlexBachmann opened the pull request:
### 1. Why is this change necessary?

It is possible that one entity is being hydrated multiple times with differently loaded associations. In these cases the current caching strategy fails to hydrate the entities correctly, because it caches the hydrated entities ONLY by its unique identifier, DISREGARDING the loaded associations of the entity. In situations, where we hydrate the same entity (same ID) two ore more times with the different associations, the entity of the first hydration is REUSED. This can lead to missing association data in the loaded entity.

### Use Case ###

Think of the following situation
```
PickupPoint
|- id
|- warehouseId
|- warehouse
|- |- id
|- |- zipcodeId
|- |- zipcode
|- |- |- id <--- ID 123
|- |- |- countryId
|- zipcodeId
|- zipcode
|- |- id <--- ID 123
|- |- countryId
|- |- country
|- |- |- id
|- |- |- iso
```

In the use case above the pickup point and the warehouse share the same zipcode (which in this use case is an entity, not a string). While the Search Query DOES NOT LOAD the country association of the zipcode entity associated with the warehouse, it does load the associated country of the zipcode entity associated with the pickup point.

The above use case is currently not hydrated correctly, since the first time the zipcode is being hydrated, the country data is missing, the country data is also missing for the second zipcode entity, since it is being reused, using the current caching strategy.

### 2. What does this change do, exactly?

The change changes the way we generate the caching key within the entity hydrator, differentiating between entities that have loaded associations and those who do not.

### 3. Describe each step to reproduce the issue or behaviour.

See supplied Unit Test

### 4. Please link to the relevant issues (if any).
https://issues.shopware.com/issues/NEXT-13735

### 5. Checklist

- [x] I have written tests and verified that they fail without my change
- [x] I have squashed any insignificant commits
- [ ] I have created a [changelog file](https://github.com/shopware/platform/blob/trunk/adr/2020-08-03-Implement-New-Changelog.md) with all necessary information about my changes
- [ ] I have written or adjusted the documentation according to my changes
- [ ] This change has comments for package types, values, functions, and non-obvious lines of code
- [x] I have read the contribution requirements and fulfil them.

bneumann97 commented:
Hey @AlexBachmann, thanks for your contribution 💙  Your `generateCacheKey` function will most probably take too much time. 
After an internal discussion, we decided that you can take out the caching here. That would be much better performance wise and also fix the issue.
And also we need a [changelog file](https://github.com/shopware/platform/blob/trunk/adr/2020-08-03-Implement-New-Changelog.md) here 🙂

AlexBachmann commented:
I took out the object caching and added a changelog file

bneumann97 commented on changelog/_unreleased/2021-02-15-remove-entity-caching-in-dbal-entity-hydrator.md:
```diff
@@ -0,0 +1,11 @@
+author_github: AlexBachmann
+---
+# Core
+* Removed entity caching in DBAL Entity Hydrator to avoid
```
I think something went wrong with your changelog here. It should only be single line not multi line. Would be great if you could change that. Besides that the PR looks good 🙂

AlexBachmann replied:
I inlined the entry

AlexBachmann committed:
(bugfix) Fix DBAL Entity Hydration when one entity is hydrated multiple times with different associations loaded

It is possible that one entity is being hydrated multiple times with differently loaded associations. In these cases the current caching strategy fails to hydrate the entities correctly, because it caches the hydrated entities ONLY by its unique identifier, DISREGARDING the loaded associations of the entity. In situations, where we hydrate the same entity (same ID) two ore more times with the different associations, the entity of the first hydration is REUSED. This can lead to missing association data in the loaded entity.

** Use Case **

Think of the following situation
PickupPoint
|- id
|- warehouseId
|- warehouse
|- |- id
|- |- zipcodeId
|- |- zipcode
|- |- |- id <--- ID 123
|- |- |- countryId
|- zipcodeId
|- zipcode
|- |- id <--- ID 123
|- |- countryId
|- |- country
|- |- |- id
|- |- |- iso

In the use case above the pickup point and the warehouse share the same zipcode (which in this use case is an entity, not a string). While the Search Query DOES NOT LOAD the country association of the zipcode entity associated with the warehouse, it does load the associated country of the zipcode entity associated with the pickup point.

The above use case is currently not hydrated correctly, since the first time the zipcode is being hydrated, the country data is missing, the country data is also missing for the second zipcode entity, since it is being reused, using the current caching

bneumann97 commented:
Hey @AlexBachmann, thanks for your contribution 💙
